librz/il/core_theory_vm.c

```diff
@@ -0,0 +1,701 @@
+ * \return temp RzILVal, a pointer to the rzil value instance at index
+ */
+RZ_API RzILVal rz_il_get_val_temp(RzILVM vm, int index) {
+	RzILTemp temp = vm->temp_value_list[index];
+	if (temp->type == RZIL_TEMP_VAL) {
+		return temp->data;
+	}
+
+	RzILVal val = rz_il_new_value();
+	if (temp->type == RZIL_TEMP_BOOL) {
+		Bool b = temp->data;
+		val->data.b = b;
+		val->type = RZIL_VAR_TYPE_BOOL;
+		temp->type = RZIL_TEMP_VAL;
+		temp->data = val;
+		return val;
+	}
+
+	if (temp->type == RZIL_TEMP_BV) {
+		BitVector bv = temp->data;
+		val->data.bv = bv;
+		val->type = RZIL_VAR_TYPE_BV;
+		temp->type = RZIL_TEMP_VAL;
+		temp->data = val;
+		return val;
+	}
+
+	// other types
+	printf("TYPES CANNOT CONVERT TO RZVAL\n");
+	return NULL;
+}
+
+/**
+ * Make the temporary value as an empty one without type info
+ * This function will set the data to NULL without free it
+ * The developer should manage the data manually before empty_temp
+ * \param vm RzILVM, pointer to VM
+ * \param index, index to get temporary value
+ */
+RZ_API void rz_il_empty_temp(RzILVM vm, int index) {
+	if (index < 0) {
+		return; // do nothing
+	}
+	RzILTemp temp = vm->temp_value_list[index];
+	temp->data = NULL;
+	temp->type = RZIL_TEMP_EMPTY;
+}
+
+/**
+ * Free a temporary value
+ * \param vm, pointer to VM
+ * \param temp, pointer to temporary value
+ */
+RZ_API void rz_il_clean_temp(RzILVM vm, RzILTemp temp) {
+	RZIL_TEMP_TYPE type = temp->type;
+	switch (type) {
+	case RZIL_TEMP_VAL:
+		rz_il_free_value(temp->data);
+		break;
+	case RZIL_TEMP_BOOL:
+		rz_il_free_bool(temp->data);
+		break;
+	case RZIL_TEMP_BV:
+		rz_il_bv_free(temp->data);
+		break;
+	case RZIL_TEMP_EFF:
+		effect_free(temp->data);
+		break;
+	default:
+		break;
+	}
+
+	temp->data = NULL;
+	temp->type = RZIL_TEMP_EMPTY;
+}
+
+/**
+ * Free and clean the temporay values in VM
+ * \param vm RzILVM, pointer to VM
+ */
+RZ_API void rz_il_clean_temps(RzILVM vm) {
+	RzILTemp *temps = vm->temp_value_list;
+	for (int i = 0; i < VM_MAX_TEMP; ++i) {
+		if (temps[i]) {
+			rz_il_clean_temp(vm, temps[i]);
+		}
+	}
+}
+
+/**
+ * Make a core theory opcode vector
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist(int num, ...) {
+	va_list args;
+	RzILOp cur_op;
+	RzPVector *oplist = rz_pvector_new((RzPVectorFree)rz_il_free_op);
+
+	va_start(args, num);
+	for (int i = 0; i < num; ++i) {
+		cur_op = va_arg(args, RzILOp);
+		rz_pvector_push(oplist, cur_op);
+	}
+	va_end(args);
+
+	return oplist;
+}
+
+/**
+ * Make a core theory opcode vector, and set id for every opcode
+ * \param id ut64, set id for every opcode in this list
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist_with_id(ut64 id, int num, ...) {
```

Comment:
Wouldn't it be better to just pass an array of RzILOp?

Reply:
In the previous version I used an array, but I found I have to prepare the array (many times) and then I turned to this approach.